docs/instructions.md

```diff
@@ -0,0 +1,164 @@
+# External Instruction Definitions
+
+Exo allows users to define custom hardware instructions within their code using the `@instr` annotation.
+These user-defined instructions can be leveraged during the scheduling process to replace specific code fragments with calls to hardware-optimized instructions.
+
+## Overview
+
+- **Custom Instructions**: Define hardware-specific instructions as procedures using the `@instr` decorator.
+- **Replace**: Use the `replace` primitive to substitute code fragments with calls to these instructions.
+- **Code Generation**: Custom instructions can emit arbitrary C code, including inline assembly, with placeholders for arguments.
+
+## Defining Custom Instructions
+
+Custom instructions are defined as procedures annotated with `@instr`.
+The `@instr` decorator allows you to specify the C code to be emitted when the instruction is called.
+
+### Syntax
+
+```python
+@instr("C code")
+def instruction_name(args):
+    # Specification of the instruction's behavior
+```
+- **`@instr`**: Decorator that specifies the C code to emit. In the string provided to `@instr`, you can include placeholders wrapped in `{}`. These placeholders will be replaced with the names of the arguments when the code is compiled.
+- **`instruction_name`**: The name of your custom instruction.
+- **`args`**: Arguments to the instruction.
+- **semantics**: Semantics of the hardware instruction, written as Exo object code.
+
+### Example: Defining a Neon Load Instruction
+
+Below is an example of defining a NEON load instruction that loads four `f32` values into Neon memory.
+
+```python
+from exo import *
+
+@instr("{dst_data} = vld1q_f32(&{src_data});")
+def neon_vld_4xf32(dst: [f32][4] @ Neon, src: [f32][4] @ DRAM):
+    assert stride(src, 0) == 1
+    assert stride(dst, 0) == 1
+
+    for i in seq(0, 4):
+        dst[i] = src[i]
+```
+
+- **`@instr("{dst_data} = vld1q_f32(&{src_data});")`**: Specifies the C code to emit when this instruction is called.
+  - `{dst_data}` and `{src_data}` are format strings that will be replaced with the actual arguments during codegen.
```

Comment:
I think we should leave this for later since we may need to do a bunch of work to fix the issues with window expressions and custom memory types that I identified. But eventually I think we should document the placeholders for `@instr` more.

Reply:
Yeah, just telling ppl to use `_data` is probably appropriate here